<a href="https://colab.research.google.com/github/ahmadd4062/flyrank-ml-internship/blob/main/work/notebooks/w04_baseline_score.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-07 — Baseline Action Score and Top-20 Review

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

In [25]:
import os
import pandas as pd
import numpy as np
from huggingface_hub import snapshot_download
from google.colab import userdata

In [26]:
HF_TOKEN = userdata.get("HF_TOKEN")

repo_path = snapshot_download(
    repo_id="FlyRank/internship-warehouse",
    repo_type="dataset",
    token=HF_TOKEN
)

print("Warehouse loaded from:")
print(repo_path)

Reconstructing (incomplete total...): |          |  0.00B /  0.00B            

Fetching 24 files:   0%|          | 0/24 [00:00<?, ?it/s]

Warehouse loaded from:
/root/.cache/huggingface/hub/datasets--FlyRank--internship-warehouse/snapshots/50cbf7c3909d07be4d1b5906b4d09e882e5acbf2


In [27]:
march_path = os.path.join(
    repo_path,
    "fact_content_daily_performance",
    "month=2026-03",
    "data_0.parquet"
)

df = pd.read_parquet(march_path)

print("Rows:", len(df))
print("Columns:", len(df.columns))
print("Date range:", df["report_date"].min(), "to", df["report_date"].max())

df.head()

Rows: 9841378
Columns: 30
Date range: 2026-03-01 to 2026-03-31


,report_date,client_hash_id,content_hash_id,client_has_gsc,client_has_ga4,gsc_data_available,ga4_data_available,gsc_impressions,gsc_clicks,gsc_sum_position,...,sessions_paid,sessions_ai,ai_chatgpt,ai_perplexity,ai_gemini,ai_copilot,ai_claude,ai_meta,ai_other,scroll_events
0,2026-03-01,client_73cda7b4e4f265ea,content_b7e512995f79d5a6,True,False,True,None,20,0,67,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2026-03-01,client_73cda7b4e4f265ea,content_05597932fe4da067,True,False,True,None,1,0,0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,2026-03-01,client_73cda7b4e4f265ea,content_7a105f548d9c6916,True,False,True,None,125,1,616,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,2026-03-01,client_73cda7b4e4f265ea,content_905aa32a0230694e,True,False,True,None,7,0,28,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,2026-03-01,client_73cda7b4e4f265ea,content_a3ea9792f793ec72,True,False,True,None,11,0,25,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [28]:
content_path = os.path.join(
    repo_path,
    "dim_content.parquet"
)

dim_content = pd.read_parquet(content_path)

print("Content dimension shape:", dim_content.shape)
print("\nColumns:")
print(dim_content.columns.tolist())

dim_content.head()

Content dimension shape: (519606, 26)

Columns:
['client_hash_id', 'content_hash_id', 'keyword_hash_id', 'url_hash_id', 'keyword_char_count', 'keyword_token_count', 'url_char_count', 'content_created_date', 'content_updated_date', 'content_type', 'search_volume', 'competition', 'competition_level', 'cpc', 'main_intent', 'backlinks', 'category_count', 'keyword_created_date', 'provider_used', 'model_used', 'char_count', 'word_count', 'last_optimized_date', 'optimization_eligible_date', 'is_published', 'is_deleted']


,client_hash_id,content_hash_id,keyword_hash_id,url_hash_id,keyword_char_count,keyword_token_count,url_char_count,content_created_date,content_updated_date,content_type,...,category_count,keyword_created_date,provider_used,model_used,char_count,word_count,last_optimized_date,optimization_eligible_date,is_published,is_deleted
0,client_04660893ae39614a,content_004de9653278b5a4,keyword_e754999ab88dd9f2,url_d6091f18cf628794,22,4,108,2026-05-30,2026-07-01,keyword article,...,3,2026-05-12,gemini-generate-content,gemini-3-flash-preview,15682.0,2555.0,None,None,True,False
1,client_04660893ae39614a,content_00dc5efae381b2ab,keyword_4329d7aede8e208b,url_3a66d2f2e36823ca,31,6,95,2026-06-12,2026-07-01,keyword article,...,4,2026-06-01,gemini-generate-content,gemini-3-flash-preview,15438.0,2430.0,None,None,True,False
2,client_04660893ae39614a,content_01410f2556c327ac,keyword_9b08047d3d2a0406,url_809eda7a7e20b3b2,22,5,82,2026-05-09,2026-07-01,keyword article,...,4,2026-05-06,gemini-generate-content,gemini-3-flash-preview,16576.0,2645.0,None,None,True,False
3,client_04660893ae39614a,content_019f27f634053ca7,keyword_e7cec7ab1804c1c2,url_5fb42bafc4399861,14,3,92,2026-06-15,2026-06-15,keyword article,...,4,2026-06-01,gemini-generate-content,gemini-3-flash-preview,15457.0,2522.0,None,None,True,False
4,client_04660893ae39614a,content_01efa71faea45dcc,keyword_56b0062a1d8b7524,url_ece0abc3e5fb75f9,24,6,98,2026-05-21,2026-06-01,keyword article,...,4,2026-05-12,gemini-generate-content,gemini-3-flash-preview,15776.0,2552.0,None,None,True,False


In [29]:
# Check how many rows exist per content page
content_counts = (
    dim_content
    .groupby("content_hash_id")
    .size()
    .describe()
)

print(content_counts)

count    519606.0
mean          1.0
std           0.0
min           1.0
25%           1.0
50%           1.0
75%           1.0
max           1.0
dtype: float64


In [30]:
# Check the date ranges available in the content dimension
print("Created date range:")
print(dim_content["content_created_date"].min(),
      "to",
      dim_content["content_created_date"].max())

print("\nUpdated date range:")
print(dim_content["content_updated_date"].min(),
      "to",
      dim_content["content_updated_date"].max())

Created date range:
2024-10-16 to 2026-07-06

Updated date range:
2024-10-28 to 2026-07-06


In [31]:
# Check whether the content dimension has duplicate content IDs
duplicate_content = dim_content["content_hash_id"].duplicated().sum()

print("Duplicate content_hash_id rows:", duplicate_content)

Duplicate content_hash_id rows: 0


In [32]:
# Aggregate March performance to one row per content page
march_pages = (
    df.groupby(["client_hash_id", "content_hash_id"], as_index=False)
      .agg(
          impressions_30d=("gsc_impressions", "sum"),
          clicks_30d=("gsc_clicks", "sum"),
          avg_position=("gsc_avg_position", "mean"),
          days_with_gsc=("gsc_data_available", "sum")
      )
)

print("Pages:", len(march_pages))
march_pages.head()

Pages: 331437


,client_hash_id,content_hash_id,impressions_30d,clicks_30d,avg_position,days_with_gsc
0,client_0797ff3a1fc9a6a5,content_004e9c4c32e88631,0,0,NaN,0
1,client_0797ff3a1fc9a6a5,content_0236ef736698e17c,0,0,NaN,0
2,client_0797ff3a1fc9a6a5,content_025f6cfd3c298870,0,0,NaN,0
3,client_0797ff3a1fc9a6a5,content_0263d5f9b7a2ecd4,1,0,9.0,1
4,client_0797ff3a1fc9a6a5,content_02752c6c1c60161f,0,0,NaN,0


In [33]:
content_cols = [
    "client_hash_id",
    "content_hash_id",
    "content_created_date",
    "content_updated_date",
    "search_volume",
    "word_count",
    "is_published",
    "is_deleted"
]

march_pages = march_pages.merge(
    dim_content[content_cols],
    on=["client_hash_id", "content_hash_id"],
    how="left"
)

print("Rows after merge:", len(march_pages))
march_pages.head()

Rows after merge: 331437


,client_hash_id,content_hash_id,impressions_30d,clicks_30d,avg_position,days_with_gsc,content_created_date,content_updated_date,search_volume,word_count,is_published,is_deleted
0,client_0797ff3a1fc9a6a5,content_004e9c4c32e88631,0,0,NaN,0,2025-10-07,2026-05-20,10.0,3935.0,True,False
1,client_0797ff3a1fc9a6a5,content_0236ef736698e17c,0,0,NaN,0,2025-10-07,2026-05-20,20.0,4335.0,True,False
2,client_0797ff3a1fc9a6a5,content_025f6cfd3c298870,0,0,NaN,0,2025-10-07,2026-05-20,10.0,3719.0,True,False
3,client_0797ff3a1fc9a6a5,content_0263d5f9b7a2ecd4,1,0,9.0,1,2025-10-07,2026-05-20,0.0,3246.0,True,False
4,client_0797ff3a1fc9a6a5,content_02752c6c1c60161f,0,0,NaN,0,2025-10-07,2026-05-20,20.0,3641.0,True,False


In [34]:
decision_date = pd.Timestamp("2026-03-31")

march_pages["content_updated_date"] = pd.to_datetime(
    march_pages["content_updated_date"]
)

# Only updates known by the March 31 decision date
march_pages["known_update_date"] = march_pages["content_updated_date"].where(
    march_pages["content_updated_date"] <= decision_date
)

march_pages["days_since_update"] = (
    decision_date - march_pages["known_update_date"]
).dt.days

print("Rows with a known update date:", march_pages["days_since_update"].notna().sum())
print("Rows without a known update date:", march_pages["days_since_update"].isna().sum())

print("\nDays since update:")
print(march_pages["days_since_update"].describe())

Rows with a known update date: 38079
Rows without a known update date: 293358

Days since update:
count    38079.000000
mean        65.360592
std         67.131240
min          7.000000
25%         34.000000
50%         34.000000
75%         34.000000
max        303.000000
Name: days_since_update, dtype: float64


In [35]:
march_pages["staleness_bucket"] = pd.cut(
    march_pages["days_since_update"],
    bins=[-1, 90, 180, 365, float("inf")],
    labels=[
        "0-90 days",
        "91-180 days",
        "181-365 days",
        "365+ days"
    ]
)

staleness_table = (
    march_pages["staleness_bucket"]
    .value_counts()
    .sort_index()
    .rename_axis("staleness_bucket")
    .reset_index(name="n")
)

staleness_table

,staleness_bucket,n
0,0-90 days,30655
1,91-180 days,3608
2,181-365 days,3816
3,365+ days,0


In [36]:
staleness_summary = (
    march_pages
    .groupby("staleness_bucket", observed=False)
    .agg(
        n=("content_hash_id", "size"),
        median_impressions=("impressions_30d", "median"),
        median_clicks=("clicks_30d", "median")
    )
    .reset_index()
)

staleness_summary

,staleness_bucket,n,median_impressions,median_clicks
0,0-90 days,30655,196.0,0.0
1,91-180 days,3608,0.0,0.0
2,181-365 days,3816,0.0,0.0
3,365+ days,0,NaN,NaN


## 1. My rule and its reason codes

Baseline rule: Prioritize published content that has been updated a long time ago and still has meaningful search visibility. The score combines content staleness and recent impressions so that older pages with stronger observed search visibility receive higher priority for a refresh review.

Reason codes:

STALE_VISIBLE — content is older and has meaningful search impressions.
STALE_LOW_VISIBILITY — content is older but has low search visibility.
RECENT_VISIBLE — content is relatively recent but has meaningful search visibility.
LOW_PRIORITY — does not meet the stronger staleness or visibility conditions.

Action: refresh_review for the highest-priority pages; lower-scoring pages remain monitor.

This is deliberately a baseline rule, not an ML model.

## 2. Build the ranked queue (writes the CSV)

*Code the score, rank everything, write work/outputs/baseline_action_score.csv.*

In [38]:
# Build the baseline ranked queue

queue = march_pages.copy()

# Fill missing values used by the rule
queue["days_since_update"] = queue["days_since_update"].clip(lower=0)
queue["impressions_30d"] = queue["impressions_30d"].fillna(0)

# Define the two baseline signals
# Staleness: older content gets a higher value
# Visibility: more recent impressions indicate stronger observed search visibility
queue["staleness_score"] = queue["days_since_update"]

queue["visibility_score"] = (
    queue["impressions_30d"].clip(upper=10000) / 100
)

# Combined baseline score
queue["score"] = (
    0.5 * queue["staleness_score"]
    + 0.5 * queue["visibility_score"]
)

# Define simple baseline conditions
stale = queue["days_since_update"] > 90
visible = queue["impressions_30d"] >= 100

# Assign exactly one reason code
queue["reason_code"] = "LOW_PRIORITY"

queue.loc[stale & visible, "reason_code"] = "STALE_VISIBLE"
queue.loc[stale & ~visible, "reason_code"] = "STALE_LOW_VISIBILITY"
queue.loc[~stale & visible, "reason_code"] = "RECENT_VISIBLE"

# Action
queue["action"] = "monitor"

# Highest-priority candidates receive refresh_review
queue.loc[
    queue["reason_code"].isin(["STALE_VISIBLE", "STALE_LOW_VISIBILITY"]),
    "action"
] = "refresh_review"

# Rank highest score first
queue = queue.sort_values(
    "score",
    ascending=False
).reset_index(drop=True)

queue["rank"] = queue.index + 1

# Create the required ranked queue
baseline_queue = queue[
    [
        "rank",
        "client_hash_id",
        "content_hash_id",
        "score",
        "reason_code",
        "action",
        "days_since_update",
        "impressions_30d",
        "clicks_30d",
        "avg_position"
    ]
].copy()

# Write the CSV required by the assignment
output_path = "work/outputs/baseline_action_score.csv"

os.makedirs("work/outputs", exist_ok=True)

baseline_queue.to_csv(output_path, index=False)

print(f"Saved: {output_path}")
print(f"Rows ranked: {len(baseline_queue):,}")

baseline_queue.head(10)

Saved: work/outputs/baseline_action_score.csv
Rows ranked: 331,437


,rank,client_hash_id,content_hash_id,score,reason_code,action,days_since_update,impressions_30d,clicks_30d,avg_position
0,1,client_c182d11e4862a37d,content_42ce26be1ec6be00,154.055,STALE_VISIBLE,refresh_review,264.0,4411,6,4.262553
1,2,client_65de48885f4ef01b,content_19f71daba0876547,151.685,STALE_LOW_VISIBILITY,refresh_review,303.0,37,0,43.347222
2,3,client_65de48885f4ef01b,content_cdd557d042d8a774,151.595,STALE_LOW_VISIBILITY,refresh_review,303.0,19,0,13.316667
3,4,client_65de48885f4ef01b,content_bb264d92c1b8bb6e,151.565,STALE_LOW_VISIBILITY,refresh_review,303.0,13,0,9.100000
4,5,client_65de48885f4ef01b,content_27705ab6c4138e8e,151.545,STALE_LOW_VISIBILITY,refresh_review,303.0,9,0,34.000000
5,6,client_65de48885f4ef01b,content_a2e85700106a33b8,151.545,STALE_LOW_VISIBILITY,refresh_review,303.0,9,0,5.566667
6,7,client_65de48885f4ef01b,content_ee6f61ff4145746c,151.535,STALE_LOW_VISIBILITY,refresh_review,303.0,7,0,26.791667
7,8,client_65de48885f4ef01b,content_300fce3631bca7e3,151.530,STALE_LOW_VISIBILITY,refresh_review,303.0,6,0,4.250000
8,9,client_65de48885f4ef01b,content_38c60323fd1608ec,151.525,STALE_LOW_VISIBILITY,refresh_review,303.0,5,0,6.625000
9,10,client_65de48885f4ef01b,content_4be96fb863be863e,151.515,STALE_LOW_VISIBILITY,refresh_review,303.0,3,0,6.333333


## 3. Top-20 review

*For each of the top 20: action, reason code, confidence note, and what would make it wrong.*

In [39]:
# Review the top 20 baseline recommendations

top20 = baseline_queue.head(20).copy()

top20["confidence_note"] = top20.apply(
    lambda row: (
        "Higher confidence because the page is stale and has meaningful "
        "observed search visibility."
        if row["reason_code"] == "STALE_VISIBLE"
        else
        "Lower confidence because the page is stale but has limited "
        "observed search visibility."
        if row["reason_code"] == "STALE_LOW_VISIBILITY"
        else
        "Moderate confidence because the page has observed visibility "
        "but is not strongly stale."
        if row["reason_code"] == "RECENT_VISIBLE"
        else
        "Low confidence because neither strong staleness nor visibility "
        "is present."
    ),
    axis=1
)

top20["what_would_make_it_wrong"] = top20.apply(
    lambda row: (
        "The page may have been intentionally left unchanged, or the "
        "observed visibility may not justify a refresh."
        if row["reason_code"] == "STALE_VISIBLE"
        else
        "Low visibility may mean a refresh is unlikely to have enough "
        "search impact to justify the effort."
        if row["reason_code"] == "STALE_LOW_VISIBILITY"
        else
        "The page may be performing adequately despite being relatively "
        "recent, so a refresh may not be necessary."
        if row["reason_code"] == "RECENT_VISIBLE"
        else
        "The available signals may miss other reasons why the page "
        "should be reviewed."
    ),
    axis=1
)

top20_review = top20[
    [
        "rank",
        "content_hash_id",
        "action",
        "reason_code",
        "score",
        "days_since_update",
        "impressions_30d",
        "confidence_note",
        "what_would_make_it_wrong"
    ]
]

top20_review

,rank,content_hash_id,action,reason_code,score,days_since_update,impressions_30d,confidence_note,what_would_make_it_wrong
0,1,content_42ce26be1ec6be00,refresh_review,STALE_VISIBLE,154.055,264.0,4411,Higher confidence because the page is stale an...,The page may have been intentionally left unch...
1,2,content_19f71daba0876547,refresh_review,STALE_LOW_VISIBILITY,151.685,303.0,37,Lower confidence because the page is stale but...,Low visibility may mean a refresh is unlikely ...
2,3,content_cdd557d042d8a774,refresh_review,STALE_LOW_VISIBILITY,151.595,303.0,19,Lower confidence because the page is stale but...,Low visibility may mean a refresh is unlikely ...
3,4,content_bb264d92c1b8bb6e,refresh_review,STALE_LOW_VISIBILITY,151.565,303.0,13,Lower confidence because the page is stale but...,Low visibility may mean a refresh is unlikely ...
4,5,content_27705ab6c4138e8e,refresh_review,STALE_LOW_VISIBILITY,151.545,303.0,9,Lower confidence because the page is stale but...,Low visibility may mean a refresh is unlikely ...
5,6,content_a2e85700106a33b8,refresh_review,STALE_LOW_VISIBILITY,151.545,303.0,9,Lower confidence because the page is stale but...,Low visibility may mean a refresh is unlikely ...
6,7,content_ee6f61ff4145746c,refresh_review,STALE_LOW_VISIBILITY,151.535,303.0,7,Lower confidence because the page is stale but...,Low visibility may mean a refresh is unlikely ...
7,8,content_300fce3631bca7e3,refresh_review,STALE_LOW_VISIBILITY,151.530,303.0,6,Lower confidence because the page is stale but...,Low visibility may mean a refresh is unlikely ...
8,9,content_38c60323fd1608ec,refresh_review,STALE_LOW_VISIBILITY,151.525,303.0,5,Lower confidence because the page is stale but...,Low visibility may mean a refresh is unlikely ...
9,10,content_4be96fb863be863e,refresh_review,STALE_LOW_VISIBILITY,151.515,303.0,3,Lower confidence because the page is stale but...,Low visibility may mean a refresh is unlikely ...


## 4. Weak picks + leakage check

*Which picks look wrong and why? Confirm no product flags or future windows leaked in.*

## Weak picks + leakage check

Several top-ranked pages are weak picks because their high score is driven mainly by staleness while recent search visibility is very low or zero. These pages are still reasonable candidates for review under the baseline rule, but confidence is low because observed visibility is weak.

The strongest pick is the page with 264 days since update and 4,411 impressions because it combines substantial staleness with meaningful observed visibility.

Leakage check:
- The score uses `days_since_update` and `impressions_30d`.
- No future-window metrics or future labels are used.
- No product flags are used as score inputs.
- `action` is an output generated by the baseline rule, not an input to the score.

In [40]:
# Leakage check

score_inputs = [
    "days_since_update",
    "impressions_30d"
]

output_columns = [
    "score",
    "reason_code",
    "action"
]

print("Score inputs:", score_inputs)

for col in score_inputs:
    print(f"{col}: present =", col in march_pages.columns)

print("\nOutput columns:", output_columns)

print("\nLeakage assessment:")
print("No future-window metrics or future labels were used.")
print("No product flags were used as score inputs.")
print("The action column is an output generated by the baseline rule, not a scoring input.")

Score inputs: ['days_since_update', 'impressions_30d']
days_since_update: present = True
impressions_30d: present = True

Output columns: ['score', 'reason_code', 'action']

Leakage assessment:
No future-window metrics or future labels were used.
No product flags were used as score inputs.
The action column is an output generated by the baseline rule, not a scoring input.


## Self-check

Before you submit, confirm each line honestly:

- [yes ] Every section above is filled — markdown thinking AND the code that backs it
- [ yes] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ yes] No client names, URLs, or private queries anywhere
- [ yes] My claims use careful words: observed, measured, directional, decision-support
- [ yes] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.